File: federift/scenario.py
"""Shared JSON scenario schema.

A *scenario* is the single source of truth shared between the Python privacy
core and the Go topology engine. Both read the same file. The Python side uses
the ``federation`` + ``privacy`` blocks; the Go engine uses the ``network``
block. Fields it doesn't recognise are ignored on each side, so the two tools
stay loosely coupled.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict


@dataclass
class Scenario:
    name: str
    seed: int
    federation: Dict[str, Any] = field(default_factory=dict)
    privacy: Dict[str, Any] = field(default_factory=dict)
    network: Dict[str, Any] = field(default_factory=dict)
    raw: Dict[str, Any] = field(default_factory=dict)

    # ---- federation getters (with defaults) --------------------------------
    @property
    def num_clients(self) -> int:
        return int(self.federation.get("num_clients", 10))

    @property
    def rounds(self) -> int:
        return int(self.federation.get("rounds", 20))

    @property
    def dim(self) -> int:
        return int(self.federation.get("dim", 16))

    @property
    def num_classes(self) -> int:
        return int(self.federation.get("num_classes", 10))

    @property
    def num_samples(self) -> int:
        return int(self.federation.get("num_samples", 5000))

    @property
    def partition(self) -> str:
        return str(self.federation.get("partition", "dirichlet"))

    @property
    def alpha(self) -> float:
        return float(self.federation.get("alpha", 0.5))

    @property
    def clients_per_round(self) -> int:
        return int(self.federation.get("clients_per_round", self.num_clients))

    @property
    def lr(self) -> float:
        return float(self.federation.get("lr", 0.5))

    @property
    def jitter(self) -> float:
        return float(self.federation.get("jitter", 0.05))

    @property
    def aggregator(self) -> str:
        return str(self.federation.get("aggregator", "fedavg"))

File: topology/scenario/scenario.go
// Package scenario parses the shared JSON scenario file. Only the fields the
// Go topology engine cares about (the "network" block plus a little federation
// metadata) are decoded; everything else is ignored, keeping the Go and Python
// tools loosely coupled around one file.
package scenario

import (
	"encoding/json"
	"fmt"
	"os"
)

// Partition describes a network split isolating a set of clients over a
// contiguous range of rounds [RoundStart, RoundEnd).
type Partition struct {
	RoundStart int   `json:"round_start"`
	RoundEnd   int   `json:"round_end"`
	Isolated   []int `json:"isolated"`
}

// Network holds all topology-engine parameters.
type Network struct {
	BaseLatencyMs    float64     `json:"base_latency_ms"`
	JitterMs         float64     `json:"jitter_ms"`
	DropProb         float64     `json:"drop_prob"`
	StragglerProb    float64     `json:"straggler_prob"`
	StragglerExtraMs float64     `json:"straggler_extra_ms"`
	DeadlineMs       float64     `json:"deadline_ms"`
	Partitions       []Partition `json:"partitions"`
}

// Federation carries just enough shape for the engine to know how many
// clients/rounds to simulate. The privacy fields live here too but are unused
// on the Go side.
type Federation struct {
	NumClients     int `json:"num_clients"`
	Rounds         int `json:"rounds"`
	ClientsPerRound int `json:"clients_per_round"`
}

// Scenario is the top-level shared document.
type Scenario struct {
	Name       string     `json:"name"`
	Seed       int64      `json:"seed"`
	Federation Federation `json:"federation"`
	Network    Network    `json:"network"`
}

// Defaults fills sensible values for anything left at zero.
func (s *Scenario) Defaults() {
	if s.Federation.NumClients == 0 {
		s.Federation.NumClients = 10
	}
	if s.Federation.Rounds == 0 {
		s.Federation.Rounds = 20
	}
	if s.Federation.ClientsPerRound == 0 {
		s.Federation.ClientsPerRound = s.Federation.NumClients
	}
